### PR TITLE
[serve] Set a hard limit for the number of scheduling tasks in the router (`50`)

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Addresses a stability issue discovered in scale testing: in large clusters with many replicas, the number of scheduling tasks can grow very large. When enough of these tasks run, it can slow down the process enough that the response deadline for the queue length message is never reached, effectively grinding the proxy (or another caller) to a halt.

This also calls `ray.cancel` on the `get_queue_state` object ref when a timeout occurs to avoid excess load on both the replica and the calling actor (which must process the RPC and callback when it finishes).

## Related issue number

Addresses #39418

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
FYI the limit of 50 was set empirically by observing the metrics during the test. Aside from the overloaded "weakest link" proxy, the others hovered around 10-30 and basically never exceeded 50.

With this change, the slowest proxy was stable at 50 (but had higher latencies than the others).

File: python/ray/serve/_private/router.py [hunk 1]
```diff
@@ -350,6 +355,11 @@ class PowerOfTwoChoicesReplicaScheduler(ReplicaScheduler):
     # received within this deadline, the replica will not be considered.
     queue_len_response_deadline_s = 0.1
 
+    # Hard limit on the maximum number of scheduling tasks to run. Having too many of
+    # these tasks can cause stability issue due to too much load on the local process
+    # and too requests in flight to fetch replicas' queue lengths.
```
Comment (shrekris-anyscale) [inline python/ray/serve/_private/router.py:360]:
```suggestion
    # and too many requests in flight to fetch replicas' queue lengths.
```

**Review by shrekris-anyscale — APPROVED** [review]

File: python/ray/serve/_private/router.py [hunk 2]
```diff
@@ -229,9 +229,14 @@ async def get_queue_state(self) -> Tuple[int, bool]:
         # NOTE(edoakes): the `get_num_ongoing_requests` method name is shared by
         # the Python and Java replica implementations. If you change it, you need to
         # change both (or introduce a branch here).
-        queue_len = await self._actor_handle.get_num_ongoing_requests.remote()
-        accepted = queue_len < self._replica_info.max_concurrent_queries
-        return queue_len, accepted
+        obj_ref = self._actor_handle.get_num_ongoing_requests.remote()
+        try:
```
Comment (sihanwang41) [inline python/ray/serve/_private/router.py:233]: you mention `gets cancelled if we don't get a response within 100ms`
For my knowledge, where is it configured?

**Review by sihanwang41 — APPROVED** [review]

File: python/ray/serve/_private/router.py [hunk 3]
```diff
@@ -350,6 +355,11 @@ class PowerOfTwoChoicesReplicaScheduler(ReplicaScheduler):
     # received within this deadline, the replica will not be considered.
     queue_len_response_deadline_s = 0.1
 
+    # Hard limit on the maximum number of scheduling tasks to run. Having too many of
+    # these tasks can cause stability issue due to too much load on the local process
+    # and too requests in flight to fetch replicas' queue lengths.
+    max_num_scheduling_tasks_hard_limit = 50
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/router.py:361]:
```suggestion
    max_num_scheduling_tasks_cap = 50
```

File: python/ray/serve/_private/router.py [hunk 4]
```diff
@@ -415,7 +425,7 @@ def curr_num_scheduling_tasks(self) -> int:
     @property
     def max_num_scheduling_tasks(self) -> int:
         """Max number of scheduling tasks to run at any time."""
-        return 2 * len(self._replicas)
+        return min(self.max_num_scheduling_tasks_hard_limit, 2 * len(self._replicas))
```
Comment (alexeykudinkin) [inline python/ray/serve/_private/router.py:428]: For my own understanding: where's 2 * replicas heuristic is coming from?

**Review by alexeykudinkin — APPROVED** [review]

File: python/ray/serve/_private/router.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serve/_private/router.py:233]: It's in `select_from_candidate_replicas` and controlled by `queue_len_response_deadline_s`

File: python/ray/serve/_private/router.py (hunk 4 above)
Comment (edoakes) [reply python/ray/serve/_private/router.py:428]: From my brain cells and playing around with some benchmarks :)

Basic thought is to at least achieve pipeline parallelism to each replica

We should definitely keep iterating on these params

**Pull request merged**